src/mame/misc/anes.cpp

```diff
@@ -38,149 +38,587 @@ namespace {
+		m_blit_val[0] = data;
+	}
+	else if (which == 2)
+	{
+		m_blit_addr[1] = offset;
+		m_blit_val[1] = data;
+	}
+	else
+		logerror("%s: Unknown blit ROM which %02X\n", machine().describe_context(), which);
+}
+
+uint8_t anes_state::blit_status_r()
+{
+	// bits 0-3 = ?
+	// bit    4 = busy
+	return 0x00;
+}
+
+void anes_state::matrix_w(uint8_t data)
+{
+	m_inp_matrix = data;
+	if (m_inp_matrix & (~0x1f))
+		logerror("%s: Unknown mux bit written %02X\n", machine().describe_context(), data);
+}
+
+uint8_t anes_state::key_r(offs_t offset)
+{
+	uint8_t data = 0xff;
+
+	// read key matrix
+	for (int i = 0; i < 5; i++)
+		if (!BIT(m_inp_matrix, i))
+			data &= m_key[offset][i]->read();
 
-	//printf("%02x%02x%02x\n",m_vram_offset[0],m_vram_offset[1],m_vram_offset[2]);
+	return data;
+}
+
+uint8_t anes_state::m1_rom_r(offs_t offset)
+{
+	if (!machine().side_effects_disabled() && m_bank_delay)
+	{
+		if (--m_bank_delay == 0)
+			m_rombank->set_entry(m_bank);
+	}
+
+	return m_maincpu->space(AS_PROGRAM).read_byte(offset);
+}
+
+void anes_state::palette_enable_w(uint8_t data)
+{
+	m_palette_enable = data;
+}
+
+void anes_state::set_color(int offset)
+{
+	uint16_t dat = (m_palette_data_msb[offset] << 8) | palette_data_lsb[offset];
+	m_palette->set_pen_color(offset, rgb_t(pal5bit(dat >> 0), pal5bit(dat >> 5), pal5bit(dat >> 10)));
+}
+
+void anes_state::palette_offset_w(uint8_t data)
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("write to palette_offset_w when not enabled %02x (enable %02x)\n", data, m_palette_enable);
+	}
+	else
+	{
+		m_palette_offset = data;
+	}
+}
+
+void anes_state::palette_offset_msb_w(uint8_t data)
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("%s: write to palette_offset_msb_w when not enabled %02x (enable %02x)\n", machine().describe_context(), data, m_palette_enable);
+	}
+	else
+	{
+		m_palette_offset_msb = data;
+	}
+}
+
+void anes_state::palette_active_bank_w(uint8_t data)
+{
+	m_palette_active_bank = data & 0xf;
+}
+
+uint8_t anes_state::palette_data2_r()
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("read from palette_data2_r when not enabled\n");
+	}
+	else
+	{
+		int offs = (m_palette_offset + (m_palette_offset_msb << 8)) & 0x1fff;
+		return palette_data_lsb[offs];
+	}
+	return 0x00;
+}
+
+
+void anes_state::palette_data2_w(uint8_t data)
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("write to palette_data2_w when not enabled %02x\n", data);
+	}
+	else
+	{
+		int offs = (m_palette_offset + (m_palette_offset_msb << 8)) & 0x1fff;
+		palette_data_lsb[offs] = data;
+		set_color(offs);
+	}
+}
+
+void anes_state::palette_data_msb_w(uint8_t data)
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("write to palette_data_msb_w when not enabled %02x\n", data);
+	}
+	else
+	{
+		int offs = (m_palette_offset + (m_palette_offset_msb << 8)) & 0x1fff;
+		m_palette_data_msb[offs] = data;
+		set_color(offs);
+	}
+}
+
+
+void anes_state::opcodes_map(address_map &map)
+{
+	map(0x0000, 0xefff).r(FUNC(anes_state::m1_rom_r));
```

Comment:
Why limit to 0xefff? That wasn't in the one I gave you.
You see, if it happens to copy a function with bankswitch to RAM, and runs it from there, the bankswitch won't work. That can't be right.